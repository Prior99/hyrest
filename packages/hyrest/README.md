# Hyrest

[![npm](https://img.shields.io/npm/v/hyrest.svg)](https://www.npmjs.com/package/hyrest)
[![Build Status](https://travis-ci.org/Prior99/hyrest.svg?branch=master)](https://travis-ci.org/Prior99/hyrest)
[![Coverage Status](https://coveralls.io/repos/github/Prior99/hyrest/badge.svg?branch=master)](https://coveralls.io/github/Prior99/hyrest?branch=master)

Hyrest is a hybrid REST framework for both the client and the server.

The idea is to define routes using decorators and use them to both serve the REST endpoint
and call them from the frontend. When developing both server and client in the same repository
or sharing a common library with all endpoints between the both, a call to a REST endoint
is transparent, type-safe and as easy calling a method.

## Usage

Please refer to the [resources](#resources) for getting you started.

## Resources

- [Tutorial](https://prior99.gitlab.io/hyrest/docs/tutorial-about)
- [Minimal example project](https://github.com/Prior99/hyrest-todo-example)
- [Documentation](https://prior99.gitlab.io/hyrest/)
- [Guide](https://prior99.gitlab.io/hyrest/docs/preamble-about/)
- [API Reference](https://prior99.gitlab.io/hyrest/api/hyrest/)
