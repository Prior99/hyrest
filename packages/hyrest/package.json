{
  "name": "hyrest",
  "version": "0.10.0",
  "description": "A hybrid REST library for the frontend and backend.",
  "keywords": [
    "hyrest",
    "hybrid",
    "http",
    "rest",
    "api",
    "promise"
  ],
  "main": "dist/index.js",
  "author": "Frederick Gnodtke",
  "license": "MIT",
  "scripts": {
    "build": "tsc",
    "clean": "rm -rf dist",
    "docs": "typedoc --mode file --out docs/ src/"
  },
  "typings": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/Prior99/hyrest"
  },
  "files": [
    "dist",
    "src",
    "LICENSE",
    "README.md"
  ],
  "devDependencies": {
    "@types/invariant": "^2.2.29",
    "typedoc": "^0.12.0",
    "typescript": "^3.0.3"
  },
  "dependencies": {
    "http-status-codes": "^1.3.0",
    "invariant": "^2.2.2",
    "isomorphic-fetch": "^2.2.1",
    "path-to-regexp": "^2.4.0",
    "reflect-metadata": "^0.1.10"
  }
}
