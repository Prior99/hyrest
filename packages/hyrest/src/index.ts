import "path-to-regexp";

export * from "./all-keys";
export * from "./answers";
export * from "./authorization";
export * from "./configure";
export * from "./controller";
export * from "./converters";
export * from "./is-browser";
export * from "./last-call";
export * from "./parameters";
export * from "./processed";
export * from "./route";
export * from "./schema-generator";
export * from "./scope";
export * from "./transform";
export * from "./types";
export * from "./validation";
export * from "./validators";

import * as dataType from "./converters";
export const DataType = dataType;

import * as http from "./answers";
export const Http = http;
