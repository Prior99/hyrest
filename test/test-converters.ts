import { FullValidator, DataType, is, length, required } from "hyrest";

[
    "10",
    "1000",
    "0",
    "1237832",
    "1",
    "-1213",
    8,
    "10.0",
    "",
    {},
    [],
    "hello",
    "0.5",
    0.3,
    "-1234.2",
    "10.4",
    10.4,
    12397891239817987123123332232,
    "10foo",
    null, // tslint:disable-line
    undefined,
].forEach(input => {
    test(`int handles "${input}" correctly`, () => {
        expect(DataType.int(input)).toMatchSnapshot();
    });
});

[
    "1.0",
    "1.1",
    "10.0",
    "10",
    "-10",
    "-9.2",
    "foobar",
    8,
    {},
    [],
    10.9,
    217319823987123987192873123,
    "10foo",
    null, // tslint:disable-line
    undefined,
].forEach(input => {
    test(`float handles "${input}" correctly`, () => {
        expect(DataType.float(input)).toMatchSnapshot();
    });
});

[
    "",
    10,
    {},
    [],
    "test",
    "10",
    null, // tslint:disable-line
    undefined,
].forEach(input => {
    test(`str handles "${input}", correctly`, () => {
        expect(DataType.str(input)).toMatchSnapshot();
    });
});

[
    "",
    10,
    {},
    [],
    {
        nested: {},
    },
    "10",
    null, // tslint:disable-line
    undefined,
].forEach(input => {
    test(`obj handles "${input}", correctly`, () => {
        expect(DataType.obj(input)).toMatchSnapshot();
    });
});

[
    {
        dataType: is(DataType.str).validate(length(1, 3)),
        tests: [
            ["hi", "yo", "1"], // Snapshot 1.
            ["hi", "yo", "1", ""], // Snapshot 2.
            [1, 2, 3, 4], // Snapshot 3.
            ["1", "2", 3], // Snapshot 4.
            [], // Snapshot 5.
        ],
    },
    {
        dataType: is(DataType.obj).schema({ value: is(DataType.int).validate(required) }),
        tests: [
            [ { value: 5 }, { value: 1 }, { value: 100} ], // Snapshot 6.
            [ {}, { value: 1 }, { value: 100} ], // Snapshot 7.
            [ { value: "test" } ], // Snapshot 8.
            [], // Snapshot 9.
            {}, // Snapshot 10.
            null, // tslint:disable-line
            undefined,
        ],
    },
].forEach(({ dataType, tests }: { dataType: FullValidator<any, any>, tests: any[][] }) => {
    tests.forEach(value => {
        test("arr with a specific datatype and input detects it as expected", async () => {
            expect(await DataType.arr(dataType)(value)).toMatchSnapshot();
        });
    });
});

test("arr with a validator and an empty array", async () => {
    expect(await DataType.arr(is(DataType.obj).schema({ test: is(DataType.str) }))([])).toMatchSnapshot();
});

test("arr with a validator and a non-empty array", async () => {
    expect(await DataType.arr(is(DataType.obj).schema({ test: is(DataType.str) }))([
        { test: "a" },
        { test: "b" },
    ])).toMatchSnapshot();
});

[
    true,
    false,
    "true",
    "false",
    "",
    10,
    {},
    [],
    {
        nested: {},
    },
    "10",
    null, // tslint:disable-line
    undefined,
].forEach(input => {
    test(`bool handles "${input}", correctly`, () => {
        expect(DataType.bool(input)).toMatchSnapshot();
    });
});
