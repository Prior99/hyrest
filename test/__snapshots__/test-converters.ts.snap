// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arr with a specific datatype and input detects it as expected 1`] = `
Processed {
  "value": Array [
    "hi",
    "yo",
    "1",
  ],
}
`;

exports[`arr with a specific datatype and input detects it as expected 2`] = `
Processed {
  "nested": Object {
    "3": Processed {
      "errors": Array [
        "Shorter than minimum length of 1.",
      ],
    },
  },
}
`;

exports[`arr with a specific datatype and input detects it as expected 3`] = `
Processed {
  "nested": Object {
    "0": Processed {
      "errors": Array [
        "Not a valid string.",
      ],
    },
    "1": Processed {
      "errors": Array [
        "Not a valid string.",
      ],
    },
    "2": Processed {
      "errors": Array [
        "Not a valid string.",
      ],
    },
    "3": Processed {
      "errors": Array [
        "Not a valid string.",
      ],
    },
  },
}
`;

exports[`arr with a specific datatype and input detects it as expected 4`] = `
Processed {
  "nested": Object {
    "2": Processed {
      "errors": Array [
        "Not a valid string.",
      ],
    },
  },
}
`;

exports[`arr with a specific datatype and input detects it as expected 5`] = `
Processed {
  "value": Array [],
}
`;

exports[`arr with a specific datatype and input detects it as expected 6`] = `
Processed {
  "value": Array [
    Object {
      "value": 5,
    },
    Object {
      "value": 1,
    },
    Object {
      "value": 100,
    },
  ],
}
`;

exports[`arr with a specific datatype and input detects it as expected 7`] = `
Processed {
  "nested": Object {
    "0": Processed {
      "nested": Object {
        "value": Processed {
          "errors": Array [
            "Missing required field.",
          ],
        },
      },
    },
  },
}
`;

exports[`arr with a specific datatype and input detects it as expected 8`] = `
Processed {
  "nested": Object {
    "0": Processed {
      "nested": Object {
        "value": Processed {
          "errors": Array [
            "Not a valid integer.",
          ],
        },
      },
    },
  },
}
`;

exports[`arr with a specific datatype and input detects it as expected 9`] = `
Processed {
  "value": Array [],
}
`;

exports[`arr with a specific datatype and input detects it as expected 10`] = `
Processed {
  "errors": Array [
    "Not an array.",
  ],
}
`;

exports[`arr with a specific datatype and input detects it as expected 11`] = `
Processed {
  "errors": Array [
    "Not an array.",
  ],
}
`;

exports[`arr with a specific datatype and input detects it as expected 12`] = `Processed {}`;

exports[`arr with a validator and a non-empty array 1`] = `
Processed {
  "value": Array [
    Object {
      "test": "a",
    },
    Object {
      "test": "b",
    },
  ],
}
`;

exports[`arr with a validator and an empty array 1`] = `
Processed {
  "value": Array [],
}
`;

exports[`bool handles "", correctly 1`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "", correctly 2`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "[object Object]", correctly 1`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "[object Object]", correctly 2`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "10", correctly 1`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "10", correctly 2`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "false", correctly 1`] = `
Object {
  "value": false,
}
`;

exports[`bool handles "false", correctly 2`] = `
Object {
  "value": false,
}
`;

exports[`bool handles "null", correctly 1`] = `
Object {
  "error": "Not a valid boolean.",
}
`;

exports[`bool handles "true", correctly 1`] = `
Object {
  "value": true,
}
`;

exports[`bool handles "true", correctly 2`] = `
Object {
  "value": true,
}
`;

exports[`bool handles "undefined", correctly 1`] = `
Object {
  "value": undefined,
}
`;

exports[`empty handles "", correctly 1`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "", correctly 2`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "[object Object]", correctly 1`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "[object Object]", correctly 2`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "10", correctly 1`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "10", correctly 2`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "false", correctly 1`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "false", correctly 2`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "null", correctly 1`] = `
Object {
  "value": null,
}
`;

exports[`empty handles "true", correctly 1`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "true", correctly 2`] = `
Object {
  "error": "Not null.",
}
`;

exports[`empty handles "undefined", correctly 1`] = `
Object {
  "error": "Not null.",
}
`;

exports[`float handles "" correctly 1`] = `
Object {
  "error": "Not a valid float.",
}
`;

exports[`float handles "[object Object]" correctly 1`] = `
Object {
  "error": "Not a valid float.",
}
`;

exports[`float handles "-9.2" correctly 1`] = `
Object {
  "value": -9.2,
}
`;

exports[`float handles "-10" correctly 1`] = `
Object {
  "value": -10,
}
`;

exports[`float handles "1.0" correctly 1`] = `
Object {
  "value": 1,
}
`;

exports[`float handles "1.1" correctly 1`] = `
Object {
  "value": 1.1,
}
`;

exports[`float handles "2.17319823987124e+26" correctly 1`] = `
Object {
  "value": 2.17319823987124e+26,
}
`;

exports[`float handles "8" correctly 1`] = `
Object {
  "value": 8,
}
`;

exports[`float handles "10" correctly 1`] = `
Object {
  "value": 10,
}
`;

exports[`float handles "10.0" correctly 1`] = `
Object {
  "value": 10,
}
`;

exports[`float handles "10.9" correctly 1`] = `
Object {
  "value": 10.9,
}
`;

exports[`float handles "10foo" correctly 1`] = `
Object {
  "value": 10,
}
`;

exports[`float handles "foobar" correctly 1`] = `
Object {
  "error": "Not a valid float.",
}
`;

exports[`float handles "null" correctly 1`] = `
Object {
  "error": "Not a valid float.",
}
`;

exports[`float handles "undefined" correctly 1`] = `
Object {
  "value": undefined,
}
`;

exports[`int handles "" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "" correctly 2`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "[object Object]" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "-1213" correctly 1`] = `
Object {
  "value": -1213,
}
`;

exports[`int handles "-1234.2" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "0" correctly 1`] = `
Object {
  "value": 0,
}
`;

exports[`int handles "0.3" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "0.5" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "1" correctly 1`] = `
Object {
  "value": 1,
}
`;

exports[`int handles "1.2397891239817987e+28" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "8" correctly 1`] = `
Object {
  "value": 8,
}
`;

exports[`int handles "10" correctly 1`] = `
Object {
  "value": 10,
}
`;

exports[`int handles "10.0" correctly 1`] = `
Object {
  "value": 10,
}
`;

exports[`int handles "10.4" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "10.4" correctly 2`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "10foo" correctly 1`] = `
Object {
  "value": 10,
}
`;

exports[`int handles "1000" correctly 1`] = `
Object {
  "value": 1000,
}
`;

exports[`int handles "1237832" correctly 1`] = `
Object {
  "value": 1237832,
}
`;

exports[`int handles "hello" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "null" correctly 1`] = `
Object {
  "error": "Not a valid integer.",
}
`;

exports[`int handles "undefined" correctly 1`] = `
Object {
  "value": undefined,
}
`;

exports[`obj handles "", correctly 1`] = `
Object {
  "error": "Not a valid object.",
}
`;

exports[`obj handles "", correctly 2`] = `
Object {
  "error": "Not a valid object.",
}
`;

exports[`obj handles "[object Object]", correctly 1`] = `
Object {
  "value": Object {},
}
`;

exports[`obj handles "[object Object]", correctly 2`] = `
Object {
  "value": Object {
    "nested": Object {},
  },
}
`;

exports[`obj handles "10", correctly 1`] = `
Object {
  "error": "Not a valid object.",
}
`;

exports[`obj handles "10", correctly 2`] = `
Object {
  "error": "Not a valid object.",
}
`;

exports[`obj handles "null", correctly 1`] = `
Object {
  "value": null,
}
`;

exports[`obj handles "undefined", correctly 1`] = `
Object {
  "value": undefined,
}
`;

exports[`str handles "", correctly 1`] = `
Object {
  "value": "",
}
`;

exports[`str handles "", correctly 2`] = `
Object {
  "error": "Not a valid string.",
}
`;

exports[`str handles "[object Object]", correctly 1`] = `
Object {
  "error": "Not a valid string.",
}
`;

exports[`str handles "10", correctly 1`] = `
Object {
  "error": "Not a valid string.",
}
`;

exports[`str handles "10", correctly 2`] = `
Object {
  "value": "10",
}
`;

exports[`str handles "null", correctly 1`] = `
Object {
  "error": "Not a valid string.",
}
`;

exports[`str handles "test", correctly 1`] = `
Object {
  "value": "test",
}
`;

exports[`str handles "undefined", correctly 1`] = `
Object {
  "value": undefined,
}
`;
